internal/pkg/docker/dockercompose/volume.go

```diff
@@ -0,0 +1,143 @@
+	"github.com/dustin/go-humanize/english"
+	"strings"
+)
+
+// The default size that Copilot reserves for a service's ephemeral on-disk storage.
+const defaultEphemeralGiB = 20
+const defaultEphemeralBytes uint64 = defaultEphemeralGiB * units.GiB
+
+type volumeConverter struct {
+	otherSvcVols  map[string][]string
+	copilotVols   map[string]*manifest.Volume
+	topLevelVols  compose.Volumes
+	tmpfsVolNames map[string]bool
+}
+
+func newVolumeConverter(topLevelVols compose.Volumes) volumeConverter {
+	return volumeConverter{
+		otherSvcVols:  map[string][]string{},
+		copilotVols:   map[string]*manifest.Volume{},
+		topLevelVols:  topLevelVols,
+		tmpfsVolNames: map[string]bool{},
+	}
+}
+
+// convertVolumes converts a list of Compose volumes into the Copilot storage equivalent.
+func (vc *volumeConverter) convertVolumes(volumes []compose.ServiceVolumeConfig, otherSvcs compose.Services) (*manifest.Storage, IgnoredKeys, error) {
+	for _, otherSvc := range otherSvcs {
+		for _, vol := range otherSvc.Volumes {
+			if vol.Type == "volume" {
+				vc.otherSvcVols[vol.Source] = append(vc.otherSvcVols[vol.Source], otherSvc.Name)
+			}
+		}
+	}
+
+	var ephemeralBytes = defaultEphemeralBytes
+	var ignored IgnoredKeys
+
+	for idx, vol := range volumes {
+		if vol.Target == "" {
+			return nil, nil, fmt.Errorf("volume mounted from %q (type %q) is missing a target mount point", vol.Source, vol.Type)
+		}
+
+		mountOpts := manifest.MountPointOpts{
+			ContainerPath: aws.String(vol.Target),
+			ReadOnly:      aws.Bool(vol.ReadOnly),
+		}
+
+		if vol.Type == "tmpfs" {
+			ephemeralBytes += uint64(vol.Tmpfs.Size)
+
+			name := fmt.Sprintf("tmpfs-%v", idx)
+
+			if vc.copilotVols[name] != nil {
+				return nil, nil, fmt.Errorf("generated tmpfs volume name %s collides with an existing volume name", name)
+			}
+
+			vc.tmpfsVolNames[name] = true
+			vc.copilotVols[name] = &manifest.Volume{
+				// efs is off by default
+				MountPointOpts: mountOpts,
+			}
+			continue
+		}
+
+		if vol.Type != "volume" {
+			// TODO (rclinard-amzn): Relax the "bind" restriction in Milestone 6
+			return nil, nil, fmt.Errorf("volume type %q is not supported yet", vol.Type)
+		}
+
+		name, ign, err := vc.checkNamedVolume(vol)
+		if err != nil {
+			return nil, nil, err
+		}
+		ignored = append(ignored, ign...)
+
+		vc.copilotVols[*name] = &manifest.Volume{
+			EFS: manifest.EFSConfigOrBool{
+				Enabled: aws.Bool(true),
+			},
+			MountPointOpts: mountOpts,
+		}
+	}
+
+	// math trick to round up to the next highest GiB if it's not an even size in GiB
+	ephemeralGiB := (ephemeralBytes + units.GiB - 1) / units.GiB
+	var storage manifest.Storage
+
+	if ephemeralGiB != defaultEphemeralGiB {
+		storage.Ephemeral = aws.Int(int(ephemeralGiB))
+	}
+
+	if len(vc.copilotVols) != 0 {
+		storage.Volumes = vc.copilotVols
+	}
+
+	return &storage, ignored, nil
+}
+
+func (vc *volumeConverter) checkNamedVolume(vol compose.ServiceVolumeConfig) (*string, IgnoredKeys, error) {
```

Comment:
why is the first return value a `*string`? if you don't need a pointer for a specific reason it should probably just be a `string`

Reply:
Good catch, fixed!